### ISSUE TITLE
[SPIRV] llama2_7b int4 f16 results are off compared to CPU

### ISSUE THREAD

**Issue by yzhang93** [iree-org/iree#14772]
### What happened?

Now we switch to test a first and second vicuna combined model. The CPU int4 results match the golden file (int4 weights in int8 container). However, on the SPIRV side the results are off, especially for the second vicuna. I suspect there's some error accumulation layer by layer.
 

### Steps to reproduce your issue

1. Download the new model: https://storage.googleapis.com/shark_tank/llama2_7b/unsharded/mlir/llama2_7b_int4.mlir

2. Build is based on [1fc815c](https://github.com/openxla/iree/commit/1fc815c474d9d48bd080b62db29be97ebc037116)

3. Compile command for SPIRV:
```
./iree-compile --iree-input-type=none --iree-vm-bytecode-module-output-format=flatbuffer-binary     --iree-hal-target-backends=vulkan --mlir-print-debuginfo     --mlir-print-op-on-diagnostic=false --iree-llvmcpu-target-cpu-features=host     --iree-stream-resource-index-bits=64 --iree-vm-target-index-bits=64     --iree-vm-bytecode-module-strip-source-map=false --iree-util-zero-fill-elided-attrs     --iree-vm-target-truncate-unsupported-floats --iree-codegen-check-ir-before-llvm-conversion=false     --iree-vulkan-target-triple=rdna3-unknown-linux --iree-opt-const-expr-hoisting=false --iree-hal-dump-executable-sources-to=hal_sources     --iree-consteval-jit-debug=true   --iree-stream-resource-max-allocation-size=3221225472 --iree-spirv-index-bits=64  ~/llama2_7b_int4.mlir -o llama2_7b_vulkan.vmfb
```

4. Compile command for CPU:
```
./iree-compile --iree-input-type=tm_tensor --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=llvm-cpu --iree-llvmcpu-target-cpu-features=host --iree-llvmcpu-target-triple=x86_64-linux-gnu --iree-llvmcpu-enable-microkernels --iree-llvmcpu-stack-allocation-limit=256000 --iree-stream-resource-index-bits=64 --iree-vm-target-index-bits=64 --iree-vm-bytecode-module-strip-source-map=true --iree-util-zero-fill-elided-attrs --iree-vm-target-truncate-unsupported-floats --iree-codegen-check-ir-before-llvm-conversion=false --iree-opt-const-expr-hoisting=False ~/llama2_7b_int4.mlir -o llama2_7b_cpu.vmfb
```

5. Run module commands for SPIRV:
```
./iree-run-module --device_allocator=caching --vulkan_vma_allocator=false --module=llama2_7b_vulkan.vmfb --device=vulkan --function=first_vicuna_forward --input=1x1xi64
```

```
./iree-run-module --device_allocator=caching --vulkan_vma_allocator=false --module=llama2_7b_vulkan.vmfb --device=vulkan --function=second_vicuna_forward --input=1x1xi64 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16  --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16 --input=1x32x1x128xf16
```

You can find the test results I have here: https://console.cloud.google.com/storage/browser/shark-public/vivian/llama2_7b_results?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&authuser=0&project=nod-cloud&prefix=&forceOnObjectsSortingFiltering=false

I've also noticed the vulkan golden results are also different from the CPU ones (I know it's definitely not the golden result :() It means there might also be some issue with int8, since it uses the int8 ops with int4 weights.

**Comment by antiagainst** [comment]
Thanks @yzhang93! A few questions:

* How are these golden results are generated? 
* Golden result differences between CPU and GPU means we have some bug somewhere else?
* Do you know at which dispatch the numbers start to diverge?

**Comment by yzhang93** [comment]
> * How are these golden results are generated?

I directly modify the int4 IR by replacing any i4 with i8. 
`sed 's/i4/i8/g' llama2_7b_int4.mlir > llama2_7b_fake_int4.mlir`
The IR used for CPU and SPIRV are the same.

> * Golden result differences between CPU and GPU means we have some bug somewhere else?

I think so. One interesting thing I noticed is the [real int4 outputs](https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_vulkan_second_results.txt) are much closer to the CPU results than the [vulkan golden outputs](https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_vulkan_second_golden.txt).

> * Do you know at which dispatch the numbers start to diverge?

It's a bit difficult to find out since both the vulkan outputs and golden outputs are off compared to CPU. I'll spend some more time to find out the diverging dispatch.

**Comment by Max191** [comment]
**[Edit] The following no longer seems to be an issue after updating my drivers**

@yzhang93 I am getting the same dispatches for cpu and vulkan, which dispatches are different for you?

I have been bisecting the model to find the first diverging dispatch, and it looks like it starts at dispatch 5 in second_vicuna_forward:

```
#map = affine_map<(d0) -> (d0)>
#map14 = affine_map<(d0) -> ()>

  flow.executable private @second_vicuna_forward_dispatch_5 {
    flow.executable.export public @second_vicuna_forward_dispatch_5_generic_4096_f32 workgroups() -> (index, index, index) {
      %x, %y, %z = flow.dispatch.workgroup_count_from_slice 
      flow.return %x, %y, %z : index, index, index
    }
    builtin.module {
      func.func @second_vicuna_forward_dispatch_5_generic_4096_f32(%arg0: !flow.dispatch.tensor<readonly:tensor<4096xf32>>, %arg1: !flow.dispatch.tensor<writeonly:tensor<f32>>) {
        %cst = arith.constant 2.000000e+00 : f32
        %cst_0 = arith.constant 0.000000e+00 : f32
        %0 = flow.dispatch.tensor.load %arg0, offsets = [0], sizes = [4096], strides = [1] : !flow.dispatch.tensor<readonly:tensor<4096xf32>> -> tensor<4096xf32>
        %1 = tensor.empty() : tensor<f32>
        %2 = linalg.fill ins(%cst_0 : f32) outs(%1 : tensor<f32>) -> tensor<f32>
        %3 = linalg.generic {indexing_maps = [#map, #map14], iterator_types = ["reduction"]} ins(%0 : tensor<4096xf32>) outs(%2 : tensor<f32>) {
        ^bb0(%in: f32, %out: f32):
          %4 = math.powf %in, %cst : f32
          %5 = arith.addf %4, %out : f32
          linalg.yield %5 : f32
        } -> tensor<f32>
        flow.dispatch.tensor.store %3, %arg1, offsets = [], sizes = [], strides = [] : tensor<f32> -> !flow.dispatch.tensor<writeonly:tensor<f32>>
        return
      }
    }
  }
```

It looks like the least significant bit of the f32 output is flipped from 0 (in CPU) to 1 (in Vulkan):
CPU (in little endian hex): `804f0734`
Vulkan (little endian hex): `814f0734`
After dispatch 5, the values continue to diverge.

I am also getting the same results from int4 and int8 versions of the model in Vulkan (int8 version being the string replaced version using the `sed` command above), so this shouldn't be a problem with i4 anymore. I think we are getting reasonable output from int4 on CPU, so I am using the output from CPU as the target output right now.

I will look further into the lowerings of this dispatch for both CPU and Vulkan

CC @antiagainst

**Comment by Max191** [comment]
Wait sorry, I think I misread above that you are getting different dispatches formed for CPU and Vulkan. You can disregard my question @yzhang93

**Comment by yzhang93** [comment]
I compared the dispatch results between vulkan int4 outputs and vulkan fake int8 outputs (with int4 weights, i.e., "vulkan golden results" in the above comments). There's some issue when using the `iree-run-module` with `-iree-flow-break-dispatch` together on CPU, so here I only compare the results on vulkan side. Here are my findings:

1. The results are the same for the first vicuna model.

2. second vicuna dispatch 4 is the first diverging dispatch. The abs(error) of dispatch 4 is
`[0.00012195 0.0001241  0.00012755 ... 0.00030649 0.00025272 0.00021404]`

3. The error aggregates, e.g., for the abs(error) of dispatch 6 is 
`[4.822e-03 7.439e-04 1.020e-03 ... 3.994e-06 4.627e-03 1.660e-03]
   [2.762e-03 1.119e-03 9.598e-03 ... 2.037e-03 1.829e-03 9.212e-04]
   [9.832e-04 6.409e-04 7.539e-04 ... 3.014e-03 2.834e-03 1.535e-03]`

But after dispatch 11 the abs(error) becomes larger
`[1.145  0.5137 0.908  ... 0.847  0.833  0.939 ]
 [0.8066 0.5244 0.799  ... 0.1895 0.2455 0.2588]
 [0.2751 0.2922 0.3645 ... 0.553  0.6504 0.4658]`

**Comment by yzhang93** [comment]
@Max191 
> I am also getting the same results from int4 and int8 versions of the model in Vulkan (int8 version being the string replaced version using the `sed` command above), so this shouldn't be a problem with i4 anymore.

We have quite different findings, maybe we should sync first before posting here

**Comment by yzhang93** [comment]
Another thing related to comparing the CPU and Vulkan results... I think the first vicuna results are also off between Vulkan and CPU. Please check my results

https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_vulkan_first_results.txt
https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_cpu_first_golden.txt

**Comment by Max191** [comment]
Hmm, yeah let's sync. I am getting similar differences between CPU and Vulkan on int4 first vicuna, but not exactly the same results.

**Comment by yzhang93** [comment]
@antiagainst we still haven't figured out why @Max191 's results are different from mine. We both fresh built on the same commit [42e54ab](https://github.com/openxla/iree/commit/42e54ab4f0bd04653096193861ab2a39a34694b2). Can you try to run the commands I posted above and check if you get the same results as what I had for vulkan?

**Comment by antiagainst** [comment]
Thanks for the due diligence, @yzhang93 and @Max191! I haven't looked into running this yet; just some quick questions regarding different numbers you have seen--are you using the same GPU? I see we have `math.powf` in the IR. I guess it's for doing square. Maybe some numeric issues due to that given it translates to the `spirv.GL.Pow` intrinsic. Wondering whether things will change if you just convert it to `x * x`. You can quickly try out by hacking in a canonicalization pattern for `math.powf`.

**Comment by yzhang93** [comment]
> are you using the same GPU? 

Yes, we are using the same GPU and have double checked on the source IR and build version etc. Now we suspect it's the AMD driver that causes the difference. I have updated my drivers and checked both on Linux and Windows, just need @Max191 's confirmation after his driver updated.

> I see we have `math.powf` in the IR. I guess it's for doing square. Maybe some numeric issues due to that given it translates to the `spirv.GL.Pow` intrinsic. Wondering whether things will change if you just convert it to `x * x`. You can quickly try out by hacking in a canonicalization pattern for `math.powf`.

Thanks for the suggestion. I'll try it tomorrow.

**Comment by Max191** [comment]
I confirmed that I had a driver issue. I have updated my drivers on Windows and I get the same results as @yzhang93 now.

**Comment by antiagainst** [comment]
Thanks! Have you cross checked another GPU? Like NVIDIA. Note that you'd need to change the target triple accordingly given that there are subgroup reduction involved.

**Comment by Max191** [comment]
> I have been bisecting the model to find the first diverging dispatch, and it looks like it starts at dispatch 5 in second_vicuna_forward:

@antiagainst @yzhang93 The discrepancy between CPU and Vulkan at dispatch 5 is gone after updating my drivers, so this may not really be affecting the correctness anymore. I am now seeing segfaults when breaking at some dispatches, which I made a new issue for here: https://github.com/openxla/iree/issues/14803

**Comment by yzhang93** [comment]
To summarize for @antiagainst and others, you can ignore the above comments except the original posting so you don't get confused.

Comparing vulkan int4 and fake int8, now the first diverging dispatch is the second vicuna dispatch_8, which is the dequantization + matvec one.

The difference is not big for this dispatch alone, the absolute error is 
`[[[4.77e-06 3.22e-06 2.38e-06 ... 3.81e-06 1.91e-06 7.15e-07]]]`

The other dispatches after this also have the error in the similar range, but then when it comes to the last dispatch the error is large
`[[1.118   1.9795  0.711   ... 1.001   0.06055 0.783  ]]`

If I disable the `FuseDequantizationMatmulPass`, then the results for vulkan int4 and vulkan fake int8 become the same, and are very close to the CPU results. Note both first and second vicuna have small differences compared to CPU.
But let's first focus to fix and make vulkan int4 results match the fake int8 results.

[vulkan int4](https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_vulkan_second_results.txt)
[vulkan fake int8](https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_vulkan_second_golden.txt)
[vulkan int4 with FuseDequantizationMatmulPass disabled ](https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_vulkan_second_results_wo_fuse.txt)
[vulkan fake int8 with FuseDequantizationMatmulPass disabled ](https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_vulkan_second_golden_wo_fuse.txt)
[cpu int4](https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_cpu_second_results.txt)
[cpu fake int8](https://storage.googleapis.com/shark-public/vivian/llama2_7b_results/llama2_7b_cpu_second_golden.txt)

**Comment by antiagainst** [comment]
> Comparing vulkan int4 and fake int8, now the first diverging dispatch is the second vicuna dispatch_8, which is the dequantization + matvec one.
> 
> The difference is not big for this dispatch alone, the absolute error is `[[[4.77e-06 3.22e-06 2.38e-06 ... 3.81e-06 1.91e-06 7.15e-07]]]`

These are fairly small differences and looks in reasonable precision rounding range.

> 
> The other dispatches after this also have the error in the similar range, but then when it comes to the last dispatch the error is large `[[1.118 1.9795 0.711 ... 1.001 0.06055 0.783 ]]`

So that means the last dispatch is kinda problematic? What's that dispatch?

**Comment by Max191** [comment]
@antiagainst After gathering some more data on the divergence of the dispatches, it seems like the divergence happens in the quantized matmul dispatches. We have been looking at the dump for an isolated quantized matmul in f32 vs f16, as well as i4 vs i8, but we have not found anything suspicious yet. It would be helpful to get some input from you since we don't have as much experience with SPIRV.

Here are the dispatches and the dumps for each:
[quantized_matmul_dispatches.zip](https://github.com/openxla/iree/files/12442578/quantized_matmul_dispatches.zip)

[dump for f16 and i4](https://drive.google.com/file/d/1tn9hACA6HbwtMes6YLuz42Xq5-XHa2Z2/view?usp=sharing)
[dump for f16 and i8](https://drive.google.com/file/d/1WW5aKMwbbox20NTyB91rj72-mNSB4cUm/view?usp=sharing)
[dump for f32 and i4](https://drive.google.com/file/d/1UfC4R40MQMtiNByk1W_imsoy0VqAtv-Z/view?usp=sharing)
[dump for f32 and i8](https://drive.google.com/file/d/1MBLRAUn7g0Bz-iet3femHjUJUmxH3xKi/view?usp=sharing)

To recap on what we have confirmed in terms of results:

1. Vulkan with i4 and f16 has some error w.r.t. Vulkan with i8 and f16
2. Vulkan with i4 and f16 and Vulkan with i8 and f16 both have some error w.r.t. CPU in f16
3. CPU with f16 and i4 matches CPU with f16 and i8
4. For the old second vicuna model (which was in f32), CPU and Vulkan all match for i8 and i4 versions

In addition, the CPU output for the f16 llama2 7b model has been generating coherent text, so it seems to be working.

This info may hopefully help you decide which dumps to compare, but we have been comparing the dumps for f16 + i4 and f32 + i4 since it seems like the issue may be arising from the switch to f16.

**Comment by Max191** [comment]
@antiagainst @yzhang93 Ok, I think I have confirmed that the issue is with f16 in the quantized matmuls (or at least there is an issue there). There may still be something with int4 (I wasn't able to test this), but given that vicuna was working with int4 I think the real issue is with f16.

> Here are the dispatches and the dumps for each:
> [quantized_matmul_dispatches.zip](https://github.com/openxla/iree/files/12442578/quantized_matmul_dispatches.zip)

I ran the vmfbs for these dispatches (I can only do the i8 ones because I can't give i4 inputs in `iree-benchmark-module`), and got the following results:

(Error numbers are the norm of the difference between outputs)
1. `CPU vs Vulkan f16: 1.736328125`
2. `CPU vs Vulkan f32: 0.00021163190831430256`

For f32, the error seems within rounding error. For f16, there is clearly something off

**Comment by antiagainst** [comment]
I'm not 100% certain what's the problem is, but it seems there are wiredness inside AMD official drivers. I tried disabling/enabling the dequant and matmul fusion on different NVIDIA / AMD GPUs (drivers) and got the following:

This is for the int4+fp16 model. CPU result I've gotten for 5950X: [cpu-5950x.txt](https://github.com/openxla/iree/files/12448937/cpu-5950x.txt)

On Ubuntu 22.04.3 with Mesa 23.1.6 RAV driver for RX 7900 XTX, I see the exact same results enabling/disabling the dequant and matmul fusion:
[amd-gpu-linux-mesa23.1.6-wo-fusion.txt](https://github.com/openxla/iree/files/12448943/amd-gpu-linux-mesa23.1.6-wo-fusion.txt)
[amd-gpu-linux-mesa23.1.6-w-fusion.txt](https://github.com/openxla/iree/files/12448933/amd-gpu-linux-mesa23.1.6-w-fusion.txt)
The delta with CPU numbers are not big, eyeballing it, seems around 0.1 or smaller typically.

On Unbuntu 22.04.3 with NVIDIA official 530.30.02 for RTX 2070 Super, similarly the exact same results enabling/disabling the dequant and matmul fusion:
[nv-gpu-530.30.02-wo-fusion.txt](https://github.com/openxla/iree/files/12448941/nv-gpu-530.30.02-wo-fusion.txt)
[nv-gpu-530.30.02-w-fusion.txt](https://github.com/openxla/iree/files/12448940/nv-gpu-530.30.02-w-fusion.txt)
The delta with CPU numbers are also not big, and seems even smaller than the above AMD numbers.

On Windows 11 with AMD official driver 23.8.1 on the _same_ RX 7900 XTX, numbers are _different_ when enabling and disabling the fusion:
[amd-gpu-win-23.8.1-wo-fusion.txt](https://github.com/openxla/iree/files/12448946/amd-gpu-win-23.8.1-wo-fusion.txt)
[amd-gpu-win-23.8.1-w-fusion.txt](https://github.com/openxla/iree/files/12448934/amd-gpu-win-23.8.1-w-fusion.txt)
And the delta is with CPU seems to be larger than the Mesa driver..

So based on the above, right now I'm afraid there might be issues in AMD drivers. We have deltas w.r.t. CPU numbers there still. I don't know what the accepted epsilon should be; it would be nice to double check whether the above actually is fine by actually embedding into the e2e follow to see if we have reasonable tokens and conversations on a NVIDIA GPU.

**Comment by antiagainst** [comment]
> Thanks! Have you cross checked another GPU? Like NVIDIA. Note that you'd need to change the target triple accordingly given that there are subgroup reduction involved.

@yzhang93 / @Max191: could you double check the Vulkan result on some NVIDIA GPUs to see if they are reasonable? Note that you'd need to change the target triple accordingly given that this involves subgroup operations.

**Comment by Max191** [comment]
I don't have an NVIDIA GPU to test on, but I am going to try to reduce our quantized matmul dispatch into a smaller spirv module that we can post as a driver bug for AMD. It seems like there is definitely something going on with the drivers

**Comment by bjacob** [comment]
Over at the other issue, https://github.com/openxla/iree/issues/14842 , I suggest that while dequantizing `i4` to `f16` was a reasonable way to get things started, a better way going forward, which would avoid all these conversations, would be to extend `i4` to `i8`, unless perhaps you are looking at some hardware target where `i8` arithmetic is not an option?

**Comment by antiagainst** [comment]
We have int4 llama2 working fine now. Nothing more actionable on this particular issue. Closing.
